packages/reporter/src/BlockListener.ts

```diff
@@ -0,0 +1,69 @@
+import { Chain } from "viem"
+import winston from "winston"
+
+import Multiclient from "./MultiClient"
+import { BlockListenerConfig } from "./types/index"
+
+class BlocksListener {
+  controllers: any[]
+  logger: winston.Logger
+  timeFetchBlocksMs: number
+  lastProcessedBlock: bigint = 0n
+  multiclient: Multiclient
+  _interval: ReturnType<typeof setInterval> | undefined // NodeJs.Timeout
+  sourceChain: Chain
+  queryBlockLength: number
+  blockBuffer: number
+
+  constructor(configs: BlockListenerConfig) {
+    this.controllers = configs.controllers
+    this.timeFetchBlocksMs = configs.timeFetchBlocksMs
+    this.logger = configs.logger
+    this.multiclient = configs.multiclient
+    this.sourceChain = configs.sourceChain
+    this.queryBlockLength = configs.queryBlockLength
+    this.blockBuffer = configs.blockBuffer
+  }
+
+  start() {
+    this._fetchBlocks()
+    this._interval = setInterval(() => {
+      this._fetchBlocks()
+    }, this.timeFetchBlocksMs)
+  }
+
+  stop() {
+    clearInterval(this._interval)
+  }
+
+  async _fetchBlocks() {
+    try {
+      const client = this.multiclient.getClientByChain(this.sourceChain)
+
+      let currentBlockNumber = await client.getBlockNumber()
+
+      if (this.queryBlockLength > 256 - this.blockBuffer) {
+        this.logger.error(`Please choose a block length less than ${256 - this.blockBuffer}!`)
+      }
+
+      const startBlock = currentBlockNumber - BigInt(this.queryBlockLength)
+      const endBlock = currentBlockNumber - BigInt(this.blockBuffer)
+      const blocks = await Promise.all(
+        Array.from(
+          { length: Number(this.queryBlockLength - this.blockBuffer + 1) },
+          (_, index) => startBlock + BigInt(index),
+        ),
+      )
+      this.logger.info(`Fetching block from ${startBlock} to ${endBlock}`)
+
+      await Promise.all(this.controllers.map((_controller: any) => _controller.onBlocks(blocks)))
+
+      this.lastProcessedBlock = endBlock
```

Comment:
Why does `this.lastProcessedBlock` is not used?

Reply:
Should the block range be: `[lastProcessedBlock+1 , currentBlockNumber]` or `[currentBlockNumber-queryBlockLength, currentBlockNumber]` ? 
I'm using the second option at the moment, it only need to check if the queryBlockLength is within the 256 limitation of block hash. If we choose the first case, I'll change the timeFetchBlocksMs in order to justify the block hash limitation. I prefer the second option because it is more flexible and we don't need to query every block.
Either way, `this.lastProcessedBlock` will show the last block number that has been processed.

Reply:
The second option is fine but if you check within the code `this.lastProcessedBlock` is not used so i don't see the reason for which we have to keep it!

And another thing:

Please move this check within the constructor:

```js
if (this.queryBlockLength > 256 - this.blockBuffer) {
        this.logger.error(`Please choose a block length less than ${256 - this.blockBuffer}!`)
      }
```

Evaluating this static condition in every loop is inefficient.